### PR TITLE
Update QuantumCircuit.barrier docstrings - Issue8076 Fix

### LINKED ISSUE #8076: Barrier API documentation is not informative
### What should we add?

A Google search for "[qiskit barrier](https://google.com/search?q=qiskit+barrier)" leads one to [its entry in the API reference](https://qiskit.org/documentation/stubs/qiskit.circuit.library.Barrier.html) as the first hit.  Unfortunately, this page contains only a circular definition, and does not even explain what a barrier does:

> Barrier instruction.
>
> Create new barrier instruction.

There is a much more useful description by @chriseclectic on StackExchange, with tens of upvotes.  I believe a similar explanation should be incorporated into the `Barrier` docstring(s).

### PR DESCRIPTION
**Pull request by DanielAja**
### Summary
Update QuantumCircuit.barrier docstrings from: 
`"""Barrier instruction."""`
to: 
`"""A barrier is a visual indicator of the grouping of a circuit section. 
    Actings as a directive for circuit compilation to separate pieces 
    of a circuit so that any optimizations or re-writes are constrained 
    to only act between barriers."""`

Based on description by @chriseclectic on [StackExchange](https://quantumcomputing.stackexchange.com/questions/8369/what-is-a-barrier-in-qiskit-circuits) as requested 

Fix #8076

### CODE REVIEW

File: qiskit/circuit/barrier.py [hunk 1]
```diff
@@ -21,7 +21,10 @@
 
 
 class Barrier(Instruction):
-    """Barrier instruction."""
+    """A barrier is a visual indicator of the grouping of a circuit section. 
+    Actings as a directive for circuit compilation to separate pieces 
+    of a circuit so that any optimizations or re-writes are constrained 
+    to only act between barriers."""
```
Comment (mtreinish) [inline qiskit/circuit/barrier.py:27]: In sphinx/docstrings the first line has special meaning it's basically a summary line that is shown in table of contents and other places. So in general it's better to have a consise top line like what was there before. I do like this description but can you do something like:

```
"""
Barrier instruction

A barrier is ...

File: qiskit/circuit/barrier.py [hunk 2]
```diff
@@ -21,7 +21,10 @@
 
 
 class Barrier(Instruction):
-    """Barrier instruction."""
+    """A barrier is a visual indicator of the grouping of a circuit section. 
+    Actings as a directive for circuit compilation to separate pieces
```
Comment (mtreinish) [inline qiskit/circuit/barrier.py:25]:
```suggestion
    It also acts as a directive for circuit compilation to separate pieces 
```

**Review by mtreinish — COMMENTED** [review]
Thanks for pushing this PR it's going to be a good documentation improvement. I just left some inline suggestions about formatting that we should make before we merge this.

File: qiskit/circuit/barrier.py (hunk 2 above)
Comment (DanielAja) [reply qiskit/circuit/barrier.py]: Thanks for the note, I have committed your suggestion!

File: qiskit/circuit/barrier.py [hunk 3]
```diff
@@ -21,7 +21,10 @@
 
 
 class Barrier(Instruction):
-    """Barrier instruction."""
+    """A barrier is a visual indicator of the grouping of a circuit section. 
+    It also acts as a directive for circuit compilation to separate pieces 
+    of a circuit so that any optimizations or re-writes are constrained 
+    to only act between barriers."""
```
Comment (1ucian0) [inline qiskit/circuit/barrier.py:27]: I would turn this around, as they main aspect of a barrier is as a directive.

Comment (DanielAja) [reply qiskit/circuit/barrier.py:27]: Makes sense, is there anything I need to do to approve this change?

Comment (1ucian0) [reply qiskit/circuit/barrier.py:27]: nope.. I think we can merge after that

File: qiskit/circuit/barrier.py [hunk 4]
```diff
@@ -21,7 +21,10 @@
 
 
 class Barrier(Instruction):
-    """Barrier instruction."""
+    """A barrier is a visual indicator of the grouping of a circuit section.
```
[the remark this answers is not in the document]
Comment (HuangJunye) [reply qiskit/circuit/barrier.py:24]: @DanielAja Thanks for your contributions? Can you please address @Cryoris comments? There is also formatting issues in the docstring. You can use `tox -eblack` to fix it. See find more details in the contributing guidelines:https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md#style-and-lint

[the remark this answers is not in the document]
Comment (DanielAja) [reply qiskit/circuit/barrier.py:24]: @Cryoris @HuangJunye I believe the comment has been resolved from a previous comment (#c7bb501 above), please let me know if there are any other changes needed to be made.

[the remark this answers is not in the document]
Comment (HuangJunye) [reply qiskit/circuit/barrier.py:24]: @DanielAja Sorry I don't see the changes are incorporated in the commit https://github.com/Qiskit/qiskit-terra/pull/9015/commits/c7bb50105648c619aed7b77df80fe3c7001916d4 Can you just add back the title "Barrier instruction" as suggested by Julien?

[the remark this answers is not in the document]
Comment (DanielAja) [reply qiskit/circuit/barrier.py]: @HuangJunye I have committed the change that @Cryoris suggested. Please let me know if there are any other changes needed to be made.

**Comment by Cryoris** [comment]
Hi @DanielAja, the lint CI job is failing, it seems you have to run black to fix it 🙂

**Review by Cryoris — APPROVED** [review]

**Pull request merged**